src/accountsdb/db.zig

```diff
@@ -4251,15 +4307,129 @@ pub const BenchmarkAccountsDB = struct {
+        // this models how some accounts are far more commonly read than others.
+        // TODO: is this distribution accurate? Probably not, but I don't have the data.
+        const pubkeys_read_weighting = try allocator.alloc(f32, n_accounts);
+        for (pubkeys_read_weighting) |*read_probability| read_probability.* = random.floatNorm(f32);
+        var indexer = try WeightedIndexer.init(allocator, pubkeys_read_weighting);
+
+        // "warm up" accounts cache
+        {
+            var i: usize = 0;
+            while (i < n_accounts) : (i += 1) {
+                const pubkey_idx = indexer.nextIndex(random);
+                _ = try accounts_db.getAccount(&pubkeys[pubkey_idx]);
+            }
+        }
+
+        // reset cache hits/misses
+        if (accounts_db.maybe_accounts_cache_mux) |*accounts_cache_mux| {
+            const accounts_cache, var accounts_cache_lg = accounts_cache_mux.writeWithLock();
+            defer accounts_cache_lg.unlock();
+            accounts_cache.cache_hits = 0;
+            accounts_cache.cache_misses = 0;
+        }
+
         var timer = try sig.time.Timer.start();
-        for (0..n_accounts) |i| {
-            const pubkey = &pubkeys[i];
-            const account = try accounts_db.getAccount(pubkey);
-            if (account.data.len != (i % 1_000)) {
-                std.debug.panic("account data len dnm {}: {} != {}", .{ i, account.data.len, (i % 1_000) });
+
+        const do_read_count = n_accounts;
+        var i: usize = 0;
+        while (i < do_read_count) : (i += 1) {
+            const pubkey_idx = indexer.nextIndex(random);
+            const account = try accounts_db.getAccount(&pubkeys[pubkey_idx]);
+            if (account.data.len != (pubkey_idx % 1_000)) {
+                std.debug.panic("account data len dnm {}: {} != {}", .{ pubkey_idx, account.data.len, (pubkey_idx % 1_000) });
             }
         }
+
         const elapsed = timer.read();
+
+        if (accounts_db.maybe_accounts_cache_mux) |*accounts_cache_mux| {
+            const accounts_cache, var accounts_cache_lg = accounts_cache_mux.writeWithLock();
+            defer accounts_cache_lg.unlock();
+
+            const cache_hits = accounts_cache.cache_hits;
+            const cache_misses = accounts_cache.cache_misses;
+
+            std.debug.print(
+                "cache hits/misses : {: >7}/{: <7} ({d:.2}%) - total cache accesses: {}\n",
+                .{
+                    cache_hits,
+                    cache_misses,
+                    @as(f32, @floatFromInt(cache_hits * 100)) / @as(f32, @floatFromInt(cache_hits + cache_misses)),
+                    cache_hits + cache_misses,
+                },
+            );
+        }
+
         return elapsed;
     }
 };
+
+/// Implementation of Alastair J. Walker's "Alias method".
+/// Once constructed, this allows for efficiently getting pseudorandom indeces
+/// that fit the given probabilities.
+const WeightedIndexer = struct {
```

Comment:
should probs be in src/rand/rand.zig

Reply:
https://github.com/Syndica/sig/pull/325/commits/85ec8c3441692383fed14ff3f66f0c71e1cd35df#diff-ad5c9043014490f7d5452d02439dcbb3bf5eb93306be2a8b12ba05f7ddad5e3aR429-R438